pumpkin-lib/src/propagators/arithmetic/maximum.rs

```diff
@@ -67,23 +69,120 @@ impl<ElementVar: IntegerVariable, Rhs: IntegerVariable> Propagator
+        let mut lb_reason = predicate![self.array[0] >= max_lb];
         for var in self.array.iter() {
+            // Rule 1.
             context.set_upper_bound(var, rhs_ub, conjunction!([self.rhs <= rhs_ub]))?;
-            lb_reason.push(predicate![var >= context.lower_bound(var)]);
-            ub_reason.push(predicate![var <= context.upper_bound(var)]);
 
-            max_ub = i32::max(context.upper_bound(var), max_ub);
-            max_lb = i32::max(context.lower_bound(var), max_lb);
+            let var_lb = context.lower_bound(var);
+            let var_ub = context.upper_bound(var);
+
+            if var_lb > max_lb {
+                max_lb = var_lb;
+                lb_reason = predicate![var >= var_lb];
+            }
+
+            if var_ub > max_ub {
+                max_ub = var_ub;
+            }
+        }
+        // Rule 2.
+        context.set_lower_bound(&self.rhs, max_lb, PropositionalConjunction::from(lb_reason))?;
+
+        // Rule 3.
+        if rhs_ub > max_ub {
+            let ub_reason: PropositionalConjunction = self
+                .array
+                .iter()
+                .map(|var| predicate![var <= max_ub])
+                .collect();
+            context.set_upper_bound(&self.rhs, max_ub, ub_reason)?;
         }
 
-        context.set_upper_bound(&self.rhs, max_ub, PropositionalConjunction::new(ub_reason))?;
-        context.set_lower_bound(&self.rhs, max_lb, PropositionalConjunction::new(lb_reason))?;
+        // Rule 4.
+        // if LB(rhs) > UB(a_i) for all i, then conflict.
+        // This rule can be skipped because the previous rules establish UB(rhs) <= max{UB(a_i)}.
+
+        // Rule 5.
+        // If there is only one variable with UB(a_i) >= LB(rhs),
+        // then the bounds for rhs and that variable should be intersected.
+        let rhs_lb = context.lower_bound(&self.rhs);
+        let mut propagating_variable: Option<&ElementVar> = None;
+        let mut propagation_reason = PropositionalConjunction::default();
+        for var in self.array.iter() {
+            if context.upper_bound(var) >= rhs_lb {
+                if propagating_variable.is_none() {
+                    propagating_variable = Some(var);
+                } else {
+                    propagating_variable = None;
+                    break;
+                }
+            } else {
+                propagation_reason.add(predicate![var <= rhs_lb - 1]);
+            }
+        }
+        // If there is exactly one variable UB(a_i) >= LB(rhs), then the propagating variable is be
+        // Some. In that case, intersect the bounds of that variable and the rhs.
+        // This means taking the stronger lower and upper bound, and applying it to both variables.
+        if let Some(propagating_variable) = propagating_variable {
+            // Constrain the lower bound.
+            let var_lb = context.lower_bound(propagating_variable);
+            match var_lb.cmp(&rhs_lb) {
```

Comment:
I think it should only be the case that we can update the lower-bound of the `propagating_variable`:
- The upper-bound of `propagating_variable` is already <= than the upper-bound of the right-hand side
- The lower-bound of `rhs` is already at least as large the maximum lower-bound of the terms (which is at least as large as the lower-bound of `propagating_variable`)
- The upper-bound of `rhs` is already at least as small as the maximum upper-bound (which is necessarily the same as the upper-bound of `propagating_variable`)

Reply:
Thank you for the detailed look, indeed you are correct. I removed the redundant parts.